Update dependencies

jarrodmillman opened the pull request:
Fix #3786

dschult commented:
I'm not sure how much this matters, but the conda forges don't have pygraphviz 1.6 yet, so they will have to get that working in order to make networkx installs work.  Maybe we should leave pygraphviz requirement as >=1.5 for now... ???

jarrodmillman commented:
I changed ``pygraphviz==1.6`` to  ``pygraphviz>=1.5, <2.0``.  That should work for both pip and conda forge.  It should also prevent installing ``pygraphviz==2.0`` when that is released.

For ``gdal``, I just removed it from ``all`` so ``pip install networkx[all]`` won't cause a problem b/c of ``gdal``.  There is no good solution for having an entry in requirements that will work on all (maybe even most?) computers.  If we use,  ``gdal>=1.10.0`` then if you have already correctly installed gdal, it should work.  But if you haven't installed it, then it will likely not work.  I always have to manually update my ``requirements/extra.txt`` file to what it needs to be on each computer I use.

Maybe we should remove ``gdal`` from the requirements file as well.  I took a quick look and couldn't find another project that includes it in its requirements file.  For instance, it isn't include in Fiona:
- https://github.com/Toblerity/Fiona/blob/master/requirements.txt
- https://github.com/Toblerity/Fiona/blob/master/requirements-dev.txt

I found this project:
- https://github.com/nextgis/pygdal

But I am not sure who uses it.  We could try doing something like what Fiona does:
- https://github.com/Toblerity/Fiona/blob/master/setup.py

But that doesn't install the GDAL library.  So I am not sure how much it helps.

@dschult Does it seem reasonable to remove it from the requirements file?

I think that will greatly reduce the number of people running into gdal installation issues.  I suspect most of the people who are running into issues installing it are running into problems b/c they just want to install all optional dependencies without having a specific need for each of them.  So, if we remove it from the automatic install options, hopefully the only people who will run into problems will be people that are specifically trying to use the GDAL functionality.  When they do, they will see that they need to install GDAL.  Moreover, I suspect people who need to use it already have it installed or know how to install it.

If this seems reasonable, I may wait until after 2.5 is released.  Removing it will involve changing our CI systems.  And I have been rethinking how we handle requirement files and our dependencies recently.  There may be a bigger overhaul coming soon.
